[deprecated]Move search index to .recovery directory when database is deleted

jiangphcn opened the pull request:
<!-- Thank you for your contribution!

     Please file this form by replacing the Markdown comments
     with your text. If a section needs no action - remove it.

     Also remember, that CouchDB uses the Review-Then-Commit (RTC) model
     of code collaboration. Positive feedback is represented +1 from committers
     and negative is a -1. The -1 also means veto, and needs to be addressed
     to proceed. Once there are no objections, the PR can be merged by a
     CouchDB committer.

     See: http://couchdb.apache.org/bylaws.html#decisions for more info. -->

## Overview
Before change, the search index directories and files were directly deleted when database was deleted. This PR provides way of moving search index to directory like `/srv/search_index/.recovery/shard/60000000-7fffffff/<dbname.ts>/e66df316792ab411705e2741bba44371` when the corresponding database was deleted and "enable_database_recovery" configuration item is set to true. This allows search index files to be re-used if database is recovered. 

<!-- Please give a short brief for the pull request,
     what problem it solves or how it makes things better. -->

## Testing recommendations
```
Running com.cloudant.clouseau.IndexCleanupServiceSpec
Running com.cloudant.clouseau.IndexCleanupServiceSpec
2017-07-28 17:12:06 clouseau.cleanup [INFO] Moving 'foo' to '.recovery/foo'
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.02 sec
Tests run: 0, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.033 sec

Results :

Tests run: 85, Failures: 0, Errors: 0, Skipped: 0

[INFO]
[INFO] --- maven-jar-plugin:2.2:jar (default-jar) @ clouseau ---
[INFO]
[INFO] --- maven-assembly-plugin:2.3:single (default) @ clouseau ---
[INFO] Reading assembly descriptor: src/main/assembly/distribution.xml
[INFO] Building zip: /Users/jiangph/couchdb/labs-clouseau3/clouseau/target/clouseau-2.10.0-SNAPSHOT.zip
[INFO] Building tar: /Users/jiangph/couchdb/labs-clouseau3/clouseau/target/clouseau-2.10.0-SNAPSHOT.tar.gz
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 24.148 s
[INFO] Finished at: 2017-07-28T17:12:10+08:00
[INFO] Final Memory: 15M/299M
[INFO] -----------------------------------------------------------------------
```
<!-- Describe how we can test your changes.
     Does it provides any behaviour that the end users
     could notice? -->

## GitHub issue number
Bugzid: 86318
<!-- If this is a significant change, please file a separate issue at:
     https://github.com/apache/couchdb/issues
     and include the number here and in commit message(s) using
     syntax like "Fixes #472" or "Fixes apache/couchdb#472".  -->

## Related Pull Requests

<!-- If your changes affects multiple components in different
     repositories please put links to those pull requests here.  -->
https://github.com/cloudant-labs/clouseau/pull/15
https://github.com/cloudant/chef-repo/pull/7456
## Checklist

- [X] Code is written and works correctly;
- [X] Changes are covered by tests;
- [ ] Documentation reflects the changes;

jiangphcn committed:
Move search index to .recovery directory when database is deleted

Bugzid: 86318

rnewson commented on src/dreyfus_index_manager.erl:
```diff
@@ -84,8 +84,13 @@ handle_call({open_error, DbName, Sig, Error}, {OpenerPid, _}, State) ->
-handle_cast({cleanup, DbName}, State) ->
-    clouseau_rpc:cleanup(DbName),
+handle_cast(Msg, State) ->
+    case Msg of
```
make multiple clauses instead.

rnewson commented on src/dreyfus_index_manager.erl:
```diff
@@ -119,7 +124,15 @@ handle_db_event(DbName, created, _St) ->
     {ok, nil};
 handle_db_event(DbName, deleted, _St) ->
-    gen_server:cast(?MODULE, {cleanup, DbName}),
+    DoRecovery = config:get_boolean("couchdb",
```
`RecoveryEnabled` would be a better name, we're not 'doing' recovery here, we're just enabling a future recovery.

jiangphcn committed:
Address Robert N's comments
 - use multiple clauses for cleanup and move
 - use RecoveryEnabled to replace DoRecovery
Bugzid: 86318

jiangphcn commented:
Thanks so much for your review, Bob. I addressed them and would you please take a look?

rnewson commented:
These changes look good but don't merge yet as this needs adjusting given the comments on the related clouseau PR.

jiangphcn committed:
Address Robert N's comment
  - Move calculation of destination directory for recovering index to
    dreyfus/erlang part
Bugzid: 86318

jiangphcn commented:
Yes, need to align with the change of clouseau. I modified codes to move calculation of destination directory for recovering index to dreyfus/erlang part. Could you please take a look? Thanks

rnewson commented on src/dreyfus_index_manager.erl:
```diff
@@ -86,6 +86,12 @@ handle_call({open_error, DbName, Sig, Error}, {OpenerPid, _}, State) ->
     clouseau_rpc:cleanup(DbName),
+    {noreply, State};
+
+handle_cast({move, DbName}, State) ->
```
This doesn't match the gen_server:cast above and so it doesn't work. When I review code I am assuming that you have run through a successful move, but that doesn't seem possible with this problem.

jiangphcn replied:
Yes, I run a successful move before having this commit. It looks that `gen_server:cast(?MODULE, {move, DbName})` matches with `handle_cast({move, DbName}, State)`.

rnewson replied:
I see it, I read it backward.

rnewson commented on src/dreyfus_index_manager.erl:
```diff
@@ -119,7 +125,15 @@ handle_db_event(DbName, created, _St) ->
+        "enable_database_recovery", false),
+    case RecoveryEnabled of
+        true ->
+            gen_server:cast(?MODULE, {move, DbName});
```
same problem here

rnewson replied:
I'm also wrong here again, sorry.

rnewson commented on src/dreyfus_index_manager.erl:
```diff
@@ -86,6 +86,12 @@ handle_call({open_error, DbName, Sig, Error}, {OpenerPid, _}, State) ->
+    {noreply, State};
+
+handle_cast({move, DbName}, State) ->
+    RecoveryDir = config:get("couchdb", "recovery_index_dir", ".recovery"),
```
the recovery dir should be specified as an absolute path. I think that needs to happen in the dbcore chef cookbook. It should be under the `[:clouseau][:index_dir]` directory.

rnewson commented:
the move Src Dest code looks fine to me. The next step is to figure out how we'll deploy the change. We need to get an absolute path from the config, so please make a chef-repo PR with that. Once that exists we can ensure all three PR's will work together before merging.

jiangphcn committed:
Use absolute path for moving search index when database is deleted
Bugzid: 86318

jiangphcn commented:
Okay, issued the third PR to make directories to be configurable:
https://github.com/cloudant/chef-repo/pull/7456

jiangphcn committed:
Move calculation of destination directory to clouseau
Bugzid: 86318

rnewson commented:
+1
